magento/pwa-studio
# [PWA-1110] Investigate render/blocking in Venia (Performance)

## Description

As a developer, I want review and address issues with blockimg/response time (LCP/TBT) to achieve higher Google Lighthouse v6 performance scores as they relate to the shopping experience. This could affect the core code, Webpack and other areas. Analysis from PWA-775 can be found here: https://wiki.corp.magento.com/display/COREENG/Lighthouse+Performance+Audit+-+v8

In-scope

Deep analysis and review of blocking/response using DevTools and other resources. Recommend and remediate issues where needed. 

Requirements

Investigate render/blocking with Venia in current state
Identify and remediate (fix) issues as appropriate; possible spike or docs
Testing and validation with Google Lighthouse

## Implementation Note
This PR covers a single render blocking point at the very top of the tree added with i18n support. Instead of waiting on the network, the request flow will now cascade in this order:
1. Immediately render before messages are even loaded. This means `defaultMessage` props will be rendered at this point.
2. While network request for current store locale is in flight, use DEFAULT_LOCALE global that is queried for at build time and pre-load that language pack.
3. Once network request comes back, load language pack for that locale

For the happy path of an {{en_US}} store, this means we now immediately render at step 1, and even when step 2/3 completes, nothing in the application needs to re-render. For alternative paths that return a different locale, the app immediately renders with placeholders, and then re-renders with correctly translated strings at step 3. The flash of english strings happens in the application today, this just renders everything much faster.

Followup tickets have been added to the backlog to address the already existing flash of default locale strings; a simple solution being, always rely on build time storeConfig and preload messages while the network request is in flight. If the build time config becomes stale, the app would still recover, while still rendering immediately. We've also discussed rendering actual placeholders (grey boxes like skeleton rendering) instead of defaultMessages.

## Related Issue





* [[PWA-1110](https://jira.corp.magento.com/browse/PWA-1110)] Investigate render/blocking in Venia (Performance)

## Acceptance



### Verification Stakeholders

### Specification



### Verification Steps


1. Go to category page and clear all cache data
2. Run Lighthouse Performance audit
3. In the performance trace, verify that the `getLocale` graphql request is no longer render blocking (requests don't wait to fire until it returns)
4. Not required, but you can also compare Performance metrics with develop.pwa-venia.com. Network latency can cause these to be difficult to compare, but TTI and TBT should be lower, and the performance score should gain a few points.
5. Verify translation still works by switching to French store

## Screenshots / Screen Captures (if appropriate)

## Checklist

* I have added tests to cover my changes, if necessary.
* I have added translations for new strings, if necessary.
* I have updated the documentation accordingly, if necessary.
